tests-e2e/state_updates_test.go

```diff
@@ -0,0 +1,251 @@
+package syncv3_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	"golang.org/x/exp/slices"
+
+	"github.com/matrix-org/complement/b"
+	"github.com/matrix-org/complement/client"
+	"github.com/matrix-org/complement/must"
+	"github.com/matrix-org/sliding-sync/sync3"
+	"github.com/matrix-org/sliding-sync/testutils/m"
+)
+
+// The purpose of this test is to check that if Alice is in the room and is using the proxy, and then
+// she leaves the room so no _proxy user_ is in the room, and events are sent whilst she is gone, that
+// upon rejoining the proxy has A) the correct latest state and B) the correct latest timeline, such that
+// the prev_batch token is correct to get earlier messages.
+func TestRejoining(t *testing.T) {
+	alice := registerNamedUser(t, "alice")
+	bob := registerNamedUser(t, "bob")
+	charlie := registerNamedUser(t, "charlie")
+	doris := registerNamedUser(t, "doris")
+
+	// public chat => shared history visibility
+	roomID := bob.MustCreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+	alice.JoinRoom(t, roomID, nil)
+	sendMessage(t, alice, roomID, "first message")
+	firstTopicEventID := sendTopic(t, bob, roomID, "First Topic")
+
+	// alice is the proxy user
+	res := alice.SlidingSync(t, sync3.Request{
+		Lists: map[string]sync3.RequestList{
+			"a": {
+				RoomSubscription: sync3.RoomSubscription{
+					TimelineLimit: 5,
+					RequiredState: [][2]string{{"*", "*"}}, // all state
+				},
+				Ranges:         sync3.SliceRanges{{0, 20}},
+				BumpEventTypes: []string{"m.room.message"},
+			},
+		},
+	})
+	m.MatchResponse(t, res, m.MatchRoomSubscriptionsStrict(
+		map[string][]m.RoomMatcher{
+			roomID: {
+				m.MatchJoinCount(2),
+				m.MatchRoomName(bob.Localpart),
+				// despite asking for 5 timeline events, we only get 1 on the initial sync so prev_batch is correct
```

Comment:
Confused by this. I would have thought that Alice's timeline at least includes her join and her message? If it did, what would stop us from giving a correct prev_batch token?

Reply:
This request is an _initial_ sync v2 request. Initial sync v2 requests always use a v2 filter of timeline limit = 1, to ensure we can set a valid `prev_batch` token. This is because if we instead did 50 events and the client asked for 5 events, we have no prev_batch at event 5,4,3,2,1 which we could pass to the user. Will clarify, as I think because I neglected to emphasise "initial" as in initial v2 req, it's not clear.